File: Week2/Day9/validPerfectSquare/python/validPerfectSquareApproach2.py
## Source : https://leetcode.com/explore/challenge/card/may-leetcoding-challenge/535/week-2-may-8th-may-14th/3324/
## Author : Shubhanshu Singh
## Date   : 10-05-2020


############################### Approach 2 using Binary Search #######################################

## Note: Do not use any built-in library function such as sqrt.

## Time Complexity: O(log(n))
## Space Complexity: O(1)


class Solution:
    def isPerfectSquare(self, num: int) -> bool:
       
        l = 1
        r = num
        
        while(l<=r):
            
            mid = int(l + (r - l)/2);
            
            if(mid * mid == num):
                return True;
            elif(mid*mid > num):
                r = mid-1;
            elif(mid*mid < num):
                l = mid+1;
        return False;
        

File: Week2/Day9/validPerfectSquare/cpp/validPerfectSquareApproach2.cpp
// Source : https://leetcode.com/explore/challenge/card/may-leetcoding-challenge/535/week-2-may-8th-may-14th/3324/
// Author : Shubhanshu Singh
// Date   : 10-05-2020


/*****************  Approach 2 using simple high school mathematics *******************************/

//Note: Do not use any built-in library function such as sqrt.

//a square number follows the pattern 1+3+5+7+.....+(2*n - 1) 

// Time Complexity: O(sqrt(n))
// Space Complexity: O(1)


class Solution {
public:
    
    bool isPerfectSquare(int num) {
        
        int x = 1;
        
        while(num>0)
        {
            num -= x;
            x += 2;
        }
        
        return num == 0;
    }
};